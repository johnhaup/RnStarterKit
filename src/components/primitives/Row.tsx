import { compact, get } from 'lodash';
import React, { ReactElement } from 'react';
import { StyleSheet, View, ViewProps, ViewStyle } from 'react-native';
import { SCREEN_WIDTH } from '../../styles/measurements';

interface Props extends ViewProps {
  children?: (ReactElement | null)[] | null;
  screenWidth?: boolean;
  between?: boolean;
  stretch?: boolean;
  flex?: boolean;
  height?: number;
  style?: ViewStyle | ViewStyle[];
}

const styleSheetProps = ['screenWidth', 'between', 'stretch', 'flex'];

export const Row = ({ children, height, style, ...rest }: Props) => {
  if (!children) {
    return null;
  }

  const dynamicStyles = styleSheetProps.reduce((a: ViewStyle[], c) => {
    const value = get(rest, c, null);
    if (value) {
      return compact([...a, get(styles, c, null)]);
    }
    return a;
  }, []);

  const heightStyle = height ? { height } : null;

  const rowStyles = compact([
    styles.container,
    ...dynamicStyles,
    heightStyle,
    style,
  ]);

  return (
    <View style={rowStyles} {...rest}>
      {children}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  screenWidth: { width: SCREEN_WIDTH },
  between: { justifyContent: 'space-between' },
  stretch: { alignSelf: 'stretch' },
  flex: { flex: 1 },
});
