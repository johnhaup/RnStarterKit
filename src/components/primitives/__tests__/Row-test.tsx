/**
 * @format
 */

import React from 'react';
import 'react-native';
import { StyleSheet, View } from 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { Row } from '../Row';

it('renders correctly', () => {
  const snap = renderer.create(
    <Row>
      <View />
      <View />
    </Row>,
  );
  expect(snap).toMatchSnapshot();
});

const styles = StyleSheet.create({
  blue: { backgroundColor: 'blue' },
  red: { backgroundColor: 'red' },
});

it('renders correctly with all props', () => {
  const snap = renderer.create(
    <Row
      screenWidth
      between
      stretch
      flex
      height={10}
      style={[styles.blue, styles.red]}>
      <View />
      <View />
    </Row>,
  );
  expect(snap).toMatchSnapshot();
});

it('return null if children are null', () => {
  const componentWithoutChildren = renderer
    .create(<Row>{null}</Row>)
    .getInstance();
  expect(componentWithoutChildren).toEqual(null);
});
