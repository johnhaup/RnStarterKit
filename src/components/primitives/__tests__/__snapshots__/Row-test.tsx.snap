// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <View />
  <View />
</View>
`;

exports[`renders correctly with all props 1`] = `
<View
  between={true}
  flex={true}
  screenWidth={true}
  stretch={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      Object {
        "width": 750,
      },
      Object {
        "justifyContent": "space-between",
      },
      Object {
        "alignSelf": "stretch",
      },
      Object {
        "flex": 1,
      },
      Object {
        "height": 10,
      },
      Array [
        Object {
          "backgroundColor": "blue",
        },
        Object {
          "backgroundColor": "red",
        },
      ],
    ]
  }
>
  <View />
  <View />
</View>
`;
